### PR TITLE
perf: render pages in parallel

### PR DESCRIPTION
**Pull request by sanjaiyan-dev**
Improving the performance by running the independent async functions in parallel ✈️.

Ref-: https://github.com/vuejs/vitepress/pull/1320

_Sorry if I made any mistakes :(_

### CODE REVIEW

**Review by Mister-Hope — CHANGES_REQUESTED** [review]
This is pretty bad, as it may blow up machine memory. Image my blog have 800+ pages, with each process taking 40MB, I will use 32GB memory.

This may heavily do negative preformance effect due to Node.js "Old Space" behavior, as memory taken greater than old space will be swapped to hardrive.

You may have a test at my blog https://github.com/Mister-Hope/Mister-Hope.github.io, and I am pretty sure you will get the process killed while rendering, and a pages around 200 may get stuck for hours.

You should test this with real project to provide a brechmark with pr like this. The SSG process may actually faster if you make a "Promise queue" and limit it to 5-20 process at same time.

Welcome to try that and show brenchamrk.

**Comment by sanjaiyan-dev** [comment]
> This is pretty bad, as it may blow up machine memory. Image my blog have 800+ pages, with each process taking 40MB, I will use 32GB memory.
> 
> This may heavily do negative preformance effect due to Node.js "Old Space" behavior, as memory taken greater than old space will be swapped to hardrive.
> 
> You may have a test at my blog https://github.com/Mister-Hope/Mister-Hope.github.io, and I am pretty sure you will get the process killed while rendering, and a site with around 200 pages may get stuck for hours.
> 
> You should test this with real project to provide a brechmark with pr like this. The SSG process may actually faster if you make a "Promise queue" and limit it to 5-20 process at same time.
> 
> Welcome to try that and show brenchamrk.

Hi,
Actually I thought node js runs the process in 4 default threads according to [doc](https://nodejs.org/api/cli.html#uv_threadpool_sizesize) and it is also used for I/O tasks. And node js runs tasks concurrently or parallel according to device's capabilities.So I thought it won't break or hang the CPU by overloading because it will run certain tasks concurrently and parallel.

Ref link (about Node js parallel and concurrency) -: https://bytearcher.com/articles/parallel-vs-concurrent


_Just my thoughts might be completely wrong_

**Comment by sanjaiyan-dev** [comment]
> > This is pretty bad, as it may blow up machine memory. Image my blog have 800+ pages, with each process taking 40MB, I will use 32GB memory.
> > 
> > This may heavily do negative preformance effect due to Node.js "Old Space" behavior, as memory taken greater than old space will be swapped to hardrive.
> > 
> > You may have a test at my blog https://github.com/Mister-Hope/Mister-Hope.github.io, and I am pretty sure you will get the process killed while rendering, and a site with around 200 pages may get stuck for hours.
> > 
> > You should test this with real project to provide a brechmark with pr like this. The SSG process may actually faster if you make a "Promise queue" and limit it to 5-20 process at same time.
> > 
> > Welcome to try that and show brenchamrk.
> 
> Hi,
> Actually I thought node js runs the process in 4 default threads according to [doc](https://nodejs.org/api/cli.html#uv_threadpool_sizesize) and it is also used for I/O tasks. And node js runs tasks concurrently or parallel according to device's capabilities.So I thought it won't break or hang the CPU by overloading because it will run certain tasks concurrently and parallel.
> 
> Ref link (about Node js parallel and concurrency) -: https://bytearcher.com/articles/parallel-vs-concurrent
> 
> 
> _Just my thoughts might be completely wrong_

So I thought this won't break the CPU since it only uses the allocated amount of cores. 🥺

**Comment by Mister-Hope** [comment]
I don't think a promise get gc before promise.all is fired. 

However, I will show you some comparisons with my blog with your changes later.

**Comment by sanjaiyan-dev** [comment]
> I don't think a promise get gc before promise.all is fired.
> 
> However, I will show you some comparisons with my blog with your changes later.

Ok, Tks 💪🙌

File: packages/bundler-webpack/src/build/build.ts [hunk 1]
```diff
@@ -95,24 +95,21 @@ export const build = async (
...
-        vueRouter,
-        renderToString,
-        ssrTemplate,
-        allFilesMeta,
-        initialFilesMeta,
-        asyncFilesMeta,
-        moduleFilesMetaMap,
-      })
-    }
+    await Promise.all(app.pages?.map((page) => await renderPage({
```
Comment (Mister-Hope) [inline packages/bundler-webpack/src/build/build.ts:101]:
```suggestion
    await Promise.all(app.pages?.map((page) => renderPage({
```

**Review by Mister-Hope — CHANGES_REQUESTED** [review]

**Review by Mister-Hope — APPROVED** [review]

**Comment by sanjaiyan-dev** [comment]
> I don't think a promise get gc before promise.all is fired. 
> 
> However, I will show you some comparisons with my blog with your changes later.

Hope it worked as expected :)

**Comment by Mister-Hope** [comment]
I already approved this. Meteorlxy is in charge to merge this and publish new version.

**Comment by sanjaiyan-dev** [comment]
> I already approved this. Meteorlxy is in charge to merge this and publish new version.

Sorry, Tks 💪🤝

**Comment by meteorlxy** [comment]
The CI failed

**Comment by sanjaiyan-dev** [comment]
> The CI failed

@meteorlxy Extremely sorry, I think it failed due to some lint issues. 

I used the GitHub web based editor to edit the code can you please help me to solve the issue by running a command:) Extremely sorry for the inconvenience.

File: packages/bundler-vite/src/build/build.ts [hunk 2]
```diff
@@ -72,21 +72,20 @@ export const build = async (
     const { app: vueApp, router: vueRouter } = await createVueApp()
     const { renderToString } = await import('vue/server-renderer')
 
+    if (spinner) spinner.text = `Starting to render pages ${chalk.magenta(app.pages?.map(vuePressPage => vuePressPage.path))}`
```
Comment (Mister-Hope) [inline packages/bundler-vite/src/build/build.ts:75]: I dislike the idea here, I would choose to remove page path, for my theme docs around 800 pages, if would be annoying with tons of path throwing to logs without even a line break.

**Review by Mister-Hope — CHANGES_REQUESTED** [review]

File: packages/bundler-vite/src/build/build.ts (hunk 2 above)
Comment (sanjaiyan-dev) [reply packages/bundler-vite/src/build/build.ts]: > I dislike the idea here, I would choose to remove page path, for my theme docs around 800 pages, if would be annoying with tons of path throwing to logs without even a line break.
```js
if (spinner) spinner.text = `Starting to render pages ${chalk.magenta(app.pages.map(vuePressPage => vuePressPage.path)) + '\n'}`
```

Like this ?

Comment (sanjaiyan-dev) [reply packages/bundler-vite/src/build/build.ts]: \n

Comment (sanjaiyan-dev) [reply packages/bundler-vite/src/build/build.ts]: > I dislike the idea here, I would choose to remove page path, for my theme docs around 800 pages, if would be annoying with tons of path throwing to logs without even a line break.

@Mister-Hope  Sorry, is current change is ok ?

**Comment by meteorlxy** [comment]
Let's just remove the spinner text updates for now.

**Pull request merged**